No Locals when debugging async method...

In e-mail, Mark M. reports:

...I've got lots of locals and such, but none of them show data in the autos or local windows during a breakpoint...

```
public IHttpActionResult GetCatalogUtilization(DateTime? from = null, DateTime? to = null)
{
        if (from == null) from = DateTime.Now - TimeSpan.FromDays(30);
        if (to == null) to = DateTime.Now + TimeSpan.FromMinutes(30);

        var fromKey = $"{(DateTime.MaxValue.Ticks - from.Value.ToUniversalTime().Ticks):x16}-ffff-z";
        var toKey = $"{(DateTime.MaxValue.Ticks - to.Value.ToUniversalTime().Ticks):x16}-0000";

        return new LambdaResult(async cancelToken =>
        {
            var utilization = new Dictionary<string, Dictionary<string, ItemUtilizationData>>();

            var exchanges = await (from exch in GetSummariesTable().CreateQuery<ExchangeSummary>()
                                   where exch.PartitionKey == "EXCHANGE"
                                   where exch.RowKey.CompareTo(fromKey) <= 0 && exch.RowKey.CompareTo(toKey) >= 0
                                   select exch).AsTableQuery().ExecuteAsync();
            //pull all exchanges and build the utilization view per catalog item
            foreach (var exInfo in exchanges.OrderBy(e=>e.LogTimestamp))
            {
                if (exInfo.ExchangeKind.StartsWith("CONSUME")) continue; //we don't care about consume
                if (exInfo.ExchangeKind.Contains("CANCELLED")) continue; //cancels didn't happen
                Dictionary<string, ItemUtilizationData> catalogItemData;
                if (exInfo.CatalogItemId == null)
                {
                    //TODO: we'll need to deal with this problem
                    continue;
                }
                if (!utilization.TryGetValue(exInfo.CatalogItemId, out catalogItemData))
                {
                    catalogItemData = new Dictionary<string, ItemUtilizationData>();
                    utilization.Add(exInfo.CatalogItemId, catalogItemData);
                }
                ItemUtilizationData itemData;
                if (!catalogItemData.TryGetValue(exInfo.Item, out itemData))
                {
                    itemData = new ItemUtilizationData();
                    catalogItemData.Add(exInfo.Item, itemData);
                }
                itemData.ProcessExchange(exInfo, from.Value);
            }

            foreach (var catalogInfo in utilization)
            {
                foreach (var itemInfo in catalogInfo.Value)
                {
                    itemInfo.Value.FinalizeWindow(to.Value);
                }
            }
            var catalog = await CatalogStorage.GetCatalogAsync();

            Func<string, IEnumerable<ItemUtilizationData>> getUtilization = catId =>
            {
                Dictionary<string, ItemUtilizationData> catalogItemData;
               if (!utilization.TryGetValue(catId, out catalogItemData))
                {
                    return Enumerable.Empty<ItemUtilizationData>();
                }
                return catalogItemData.Values;
            };

            return Ok(from catalogItem in catalog
                     let utilizationInfo = getUtilization(catalogItem.Id.ToString())
                     select new CatalogItemUtilizationDto
                     {
                          Id = catalogItem.Id,
                          DisplayName = catalogItem.DisplayName,
                          Description = catalogItem.Description,
                          Category = catalogItem.Category,
                          ImagePath = Url.Link(nameof(MachineController.GetCatalogImage), new { catalogItemId = catalogItem.Id }),
                          HoursBorrowed = utilizationInfo.Sum(d => d.BorrowedTime.TotalHours),
                          Borrows = utilizationInfo.Sum(d => d.BorrowCount),
                     });
        });
}
```

Simplified repro:
```
  using System;
  using System.Threading.Tasks;

  class Program
  {
      static void Main()
      {
          Func<Task> f = async () =>
          {
              var utilization = 42;
BP>       };
          f();
      }
  }
```

Attempt to eval "utilization"

**RESULTS**
error CS0103: The name 'utilization' does not exist in the current context

--- a/src/ExpressionEvaluator/CSharp/Source/ExpressionCompiler/CompilationContext.cs
+++ b/src/ExpressionEvaluator/CSharp/Source/ExpressionCompiler/CompilationContext.cs
@@ -619,7 +619,7 @@ namespace Microsoft.CodeAnalysis.CSharp.ExpressionEvaluator
                 BinderFlags.UncheckedRegion |
                 BinderFlags.AllowManagedAddressOf |
                 BinderFlags.AllowAwaitInUnsafeContext);
-            var hasImports = !importRecordGroups.IsDefault;
+            var hasImports = !importRecordGroups.IsDefaultOrEmpty;
             var numImportStringGroups = hasImports ? importRecordGroups.Length : 0;
             var currentStringGroup = numImportStringGroups - 1;
 

--- a/src/ExpressionEvaluator/CSharp/Source/ExpressionCompiler/SymUnmanagedReaderExtensions.cs
+++ b/src/ExpressionEvaluator/CSharp/Source/ExpressionCompiler/SymUnmanagedReaderExtensions.cs
@@ -21,56 +21,59 @@ namespace Microsoft.CodeAnalysis.CSharp.ExpressionEvaluator
             var importStringGroups = reader.GetCSharpGroupedImportStrings(methodToken, methodVersion, out externAliasStrings);
             Debug.Assert(importStringGroups.IsDefault == externAliasStrings.IsDefault);
 
-            if (importStringGroups.IsDefault)
+            ArrayBuilder<ImmutableArray<ImportRecord>> importRecordGroupBuilder = null;
+            ArrayBuilder<ExternAliasRecord> externAliasRecordBuilder = null;
+            if (!importStringGroups.IsDefault)
             {
-                return default(MethodDebugInfo);
-            }
-
-            var importRecordGroupBuilder = ArrayBuilder<ImmutableArray<ImportRecord>>.GetInstance(importStringGroups.Length);
-            foreach (var importStringGroup in importStringGroups)
-            {
-                var groupBuilder = ArrayBuilder<ImportRecord>.GetInstance(importStringGroup.Length);
-                foreach (var importString in importStringGroup)
+                importRecordGroupBuilder = ArrayBuilder<ImmutableArray<ImportRecord>>.GetInstance(importStringGroups.Length);
+                foreach (var importStringGroup in importStringGroups)
                 {
-                    ImportRecord record;
-                    if (NativeImportRecord.TryCreateFromCSharpImportString(importString, out record))
+                    var groupBuilder = ArrayBuilder<ImportRecord>.GetInstance(importStringGroup.Length);
+                    foreach (var importString in importStringGroup)
                     {
-                        groupBuilder.Add(record);
+                        ImportRecord record;
+                        if (NativeImportRecord.TryCreateFromCSharpImportString(importString, out record))
+                        {
+                            groupBuilder.Add(record);
+                        }
+                        else
+                        {
+                            Debug.WriteLine($"Failed to parse import string {importString}");
+                        }
                     }
-                    else
+                    importRecordGroupBuilder.Add(groupBuilder.ToImmutableAndFree());
+                }
+
+                if (!externAliasStrings.IsDefault)
+                {
+                    externAliasRecordBuilder = ArrayBuilder<ExternAliasRecord>.GetInstance(externAliasStrings.Length);
+                    foreach (string externAliasString in externAliasStrings)
                     {
-                        Debug.WriteLine($"Failed to parse import string {importString}");
+                        string alias;
+                        string externAlias;
+                        string target;
+                        ImportTargetKind kind;
+                        if (!CustomDebugInfoReader.TryParseCSharpImportString(externAliasString, out alias, out externAlias, out target, out kind))
+                        {
+                            Debug.WriteLine($"Unable to parse extern alias '{externAliasString}'");
+                            continue;
+                        }
+
+                        Debug.Assert(kind == ImportTargetKind.Assembly, "Programmer error: How did a non-assembly get in the extern alias list?");
+                        Debug.Assert(alias != null); // Name of the extern alias.
+                        Debug.Assert(externAlias == null); // Not used.
+                        Debug.Assert(target != null); // Name of the target assembly.
+
+                        AssemblyIdentity targetIdentity;
+                        if (!AssemblyIdentity.TryParseDisplayName(target, out targetIdentity))
+                        {
+                            Debug.WriteLine($"Unable to parse target of extern alias '{externAliasString}'");
+                            continue;
+                        }
+
+                        externAliasRecordBuilder.Add(new NativeExternAliasRecord<AssemblySymbol>(alias, targetIdentity));
                     }
                 }
-                importRecordGroupBuilder.Add(groupBuilder.ToImmutableAndFree());
-            }
-
-            var externAliasRecordBuilder = ArrayBuilder<ExternAliasRecord>.GetInstance(externAliasStrings.Length);
-            foreach (string externAliasString in externAliasStrings)
-            {
-                string alias;
-                string externAlias;
-                string target;
-                ImportTargetKind kind;
-                if (!CustomDebugInfoReader.TryParseCSharpImportString(externAliasString, out alias, out externAlias, out target, out kind))
-                {
-                    Debug.WriteLine($"Unable to parse extern alias '{externAliasString}'");
-                    continue;
-                }
-
-                Debug.Assert(kind == ImportTargetKind.Assembly, "Programmer error: How did a non-assembly get in the extern alias list?");
-                Debug.Assert(alias != null); // Name of the extern alias.
-                Debug.Assert(externAlias == null); // Not used.
-                Debug.Assert(target != null); // Name of the target assembly.
-
-                AssemblyIdentity targetIdentity;
-                if (!AssemblyIdentity.TryParseDisplayName(target, out targetIdentity))
-                {
-                    Debug.WriteLine($"Unable to parse target of extern alias '{externAliasString}'");
-                    continue;
-                }
-
-                externAliasRecordBuilder.Add(new NativeExternAliasRecord<AssemblySymbol>(alias, targetIdentity));
             }
 
             var hoistedLocalScopeRecords = ImmutableArray<HoistedLocalScopeRecord>.Empty;
@@ -97,11 +100,10 @@ namespace Microsoft.CodeAnalysis.CSharp.ExpressionEvaluator
                     out dynamicLocalConstantMap);
             }
 
-
             return new MethodDebugInfo(
                 hoistedLocalScopeRecords,
-                importRecordGroupBuilder.ToImmutableAndFree(),
-                externAliasRecordBuilder.ToImmutableAndFree(),
+                importRecordGroupBuilder?.ToImmutableAndFree() ?? ImmutableArray<ImmutableArray<ImportRecord>>.Empty,
+                externAliasRecordBuilder?.ToImmutableAndFree() ?? ImmutableArray<ExternAliasRecord>.Empty,
                 dynamicLocalMap,
                 dynamicLocalConstantMap,
                 defaultNamespaceName: ""); // Unused in C#.

--- a/src/ExpressionEvaluator/CSharp/Test/ExpressionCompiler/LocalsTests.cs
+++ b/src/ExpressionEvaluator/CSharp/Test/ExpressionCompiler/LocalsTests.cs
@@ -1588,6 +1588,54 @@ class C
             locals.Free();
         }
 
+        [WorkItem(2240)]
+        [Fact]
+        public void AsyncLambda()
+        {
+            var source =
+@"using System;
+using System.Threading.Tasks;
+class C
+{
+    static void M()
+    {
+        Func<int, Task> f = async (x) =>
+        {
+            var y = 42;
+        };
+    }
+}";
+            var compilation = CreateCompilationWithMscorlib45(source, options: TestOptions.DebugDll);
+            var runtime = CreateRuntimeInstance(compilation);
+            var context = CreateMethodContext(runtime, methodName: "C.<>c.<<M>b__0_0>d.MoveNext");
+            var locals = ArrayBuilder<LocalAndMethod>.GetInstance();
+            string typeName;
+            var testData = new CompilationTestData();
+            context.CompileGetLocals(locals, argumentsOnly: false, typeName: out typeName, testData: testData);
+            Assert.Equal(locals.Count, 2);
+            VerifyLocal(testData, "<>x", locals[0], "<>m0", "x", expectedILOpt:
+@"{
+  // Code size        7 (0x7)
+  .maxstack  1
+  .locals init (int V_0,
+                System.Exception V_1)
+  IL_0000:  ldarg.0
+  IL_0001:  ldfld      ""int C.<>c.<<M>b__0_0>d.x""
+  IL_0006:  ret
+}");
+            VerifyLocal(testData, "<>x", locals[1], "<>m1", "y", expectedILOpt:
+@"{
+  // Code size        7 (0x7)
+  .maxstack  1
+  .locals init (int V_0,
+                System.Exception V_1)
+  IL_0000:  ldarg.0
+  IL_0001:  ldfld      ""int C.<>c.<<M>b__0_0>d.<y>5__1""
+  IL_0006:  ret
+}");
+            locals.Free();
+        }        
+
         [WorkItem(996571)]
         [Fact]
         public void MissingReference()

--- a/src/ExpressionEvaluator/VisualBasic/Test/ExpressionCompiler/LocalsTests.vb
+++ b/src/ExpressionEvaluator/VisualBasic/Test/ExpressionCompiler/LocalsTests.vb
@@ -1647,6 +1647,58 @@ End Class"
             locals.Free()
         End Sub
 
+        <WorkItem(2240)>
+        <Fact>
+        Public Sub AsyncLambda()
+            Const source =
+"Imports System
+Imports System.Threading.Tasks
+Class C
+    Shared Sub M()
+        Dim f As Func(Of Integer, Task) = Async Function (x)
+            Dim y = 42
+        End Function
+    End Sub
+End Class"
+            Dim comp = CreateCompilationWithReferences(
+                MakeSources(source),
+                {MscorlibRef_v4_0_30316_17626, MsvbRef_v4_0_30319_17929, SystemCoreRef_v4_0_30319_17929},
+                TestOptions.DebugDll)
+
+            Dim runtime = CreateRuntimeInstance(comp)
+            Dim context = CreateMethodContext(
+                runtime,
+                methodName:="C._Closure$__.VB$StateMachine___Lambda$__1-0.MoveNext")
+            Dim testData As New CompilationTestData()
+            Dim locals = ArrayBuilder(Of LocalAndMethod).GetInstance()
+            Dim typeName As String = Nothing
+            context.CompileGetLocals(locals, argumentsOnly:=False, typeName:=typeName, testData:=testData)
+            Assert.Equal(2, locals.Count)
+            VerifyLocal(testData, typeName, locals(0), "<>m0", "x", expectedILOpt:=
+"{
+  // Code size        7 (0x7)
+  .maxstack  1
+  .locals init (Integer V_0,
+                System.Threading.Tasks.Task V_1,
+                System.Exception V_2)
+  IL_0000:  ldarg.0
+  IL_0001:  ldfld      ""C._Closure$__.VB$StateMachine___Lambda$__1-0.$VB$Local_x As Integer""
+  IL_0006:  ret
+}")
+            VerifyLocal(testData, typeName, locals(1), "<>m1", "y", expectedILOpt:=
+"{
+  // Code size        7 (0x7)
+  .maxstack  1
+  .locals init (Integer V_0,
+                System.Threading.Tasks.Task V_1,
+                System.Exception V_2)
+  IL_0000:  ldarg.0
+  IL_0001:  ldfld      ""C._Closure$__.VB$StateMachine___Lambda$__1-0.$VB$ResumableLocal_y$0 As Integer""
+  IL_0006:  ret
+}")
+            locals.Free()
+        End Sub
+
         <WorkItem(996571)>
         <Fact>
         Public Sub MissingReference()

